src/kakarot/instructions/system_operations.cairo

```diff
@@ -581,15 +573,23 @@ namespace CallHelper {
         let stack = Stack.push(ctx.stack, status);
         let ctx = ExecutionContext.update_stack(ctx, stack);
 
-        // ret_offset, see prepare_args
-        let memory = Memory.store_n(
-            ctx.memory,
-            ctx.sub_context.return_data_len,
-            ctx.sub_context.return_data,
-            [ctx.sub_context.return_data - 1],
+        // see prepare_args
+        let ret_offset = [ctx.return_data];
```

Comment:
I wonder why we store the `ret_offset` in the first element of the return data,
It is known at the time the CALL is made, meaning that it could be a call argument just like `return_data_len`

Reply:
I don't understand this. When a CALL is made, we create a sub context, which is run in place of the calling context. When the sub context is stopped (possibly after hundreds of opcodes called), then we go back to the calling context. How could we pass this info to have it in the finalize?

Reply:
I think that I get it: we could remove the link to the `child_context` in the `ExecutionContext` and stores in the ctx.return_data the return_data of the last call. Works for me, it will save steps

Reply:
> I think that I get it: we could remove the link to the `child_context` in the `ExecutionContext` and stores in the ctx.return_data the return_data of the last call. Works for me, it will save steps

afaik this is an optimisation that is not related to my point

i meant it as a question: would it be cleaner to have `ret_offset` be its own variable, rather than being collocated at the first index of return_data

Reply:
ok so I did the non requested optim, still don't understand what you mean by "be its own variable". There is no difference with `ret_offset` and `ret_size`, both could be put in the `return_data` segment. `ret_size` is **not** the `return_data_len`.

The pb here is that we need to access at the end of the sub context data retrieved in the beginning of the sub context, and with the current implementation there is no way to easily pass these args. So we need to store them somewhere. It could be in the stack, in the memory, or as in the current implementation in the `return_data` (somehow hacky because it's not actually the return data but the args concerning the handling of the `return_data`).

The stack may make more sense, but would be less efficient

Reply:
ok, actually the CALL args are such that `ret_offset` and `ret_size` are the last items popped, so I'll update to pop them in the `finalize_calling_context` instead of directly in the `prepared_args`